Mathlib/Probability/Kernel/IonescuTulcea/Traj.lean

```diff
@@ -0,0 +1,717 @@
+of the trajectory up to time `b`, and composing this with `traj κ b` gives the distribution
+of the whole trajectory. -/
+theorem traj_comp_partialTraj {a b : ℕ} (hab : a ≤ b) :
+    (traj κ b) ∘ₖ (partialTraj κ a b) = traj κ a := by
+  refine eq_traj _ _ fun n ↦ ?_
+  rw [map_comp, traj_map_frestrictLe, partialTraj_comp_partialTraj' _ hab]
+
+/-- This theorem shows that `traj κ n` is, up to an equivalence, the product of
+a determinstic kernel with another kernel. This is an intermediate result to compute integrals
+with respect to this kernel. -/
+theorem traj_eq_prod (a : ℕ) :
+    traj κ a = (Kernel.id ×ₖ (traj κ a).map (Set.Ioi a).restrict).map (IicProdIoi a) := by
+  refine (eq_traj' _ (a + 1) _ fun b hb ↦ ?_).symm
+  rw [← map_comp_right]
+  conv_lhs => enter [2]; change (IicProdIoc a b) ∘
+    (Prod.map id (fun x i ↦ x ⟨i.1, Set.mem_Ioi.2 (mem_Ioc.1 i.2).1⟩))
+  · rw [map_comp_right, ← map_prod_map, ← map_comp_right]
+    · conv_lhs => enter [1, 2, 2]; change (Ioc a b).restrict
+      rw [← restrict₂_comp_restrict Ioc_subset_Iic_self, ← frestrictLe, map_comp_right,
+        traj_map_frestrictLe, map_id, ← partialTraj_eq_prod]
+      all_goals fun_prop
+    all_goals fun_prop
+  all_goals fun_prop
+
+theorem traj_map_updateFinset {n : ℕ} (x : Π i : Iic n, X i) :
+    (traj κ n x).map (updateFinset · _ x) = traj κ n x := by
+  nth_rw 2 [traj_eq_prod]
+  have : (updateFinset · _ x) = IicProdIoi n ∘ (Prod.mk x) ∘ (Set.Ioi n).restrict := by
+    ext; simp [IicProdIoi, updateFinset]
+  rw [this, ← Function.comp_assoc, ← Measure.map_map, ← Measure.map_map, map_apply, prod_apply,
+    map_apply, id_apply, Measure.dirac_prod]
+  all_goals fun_prop
+
+end basic
+
+section integral
+
+/-! ### Integrals and `traj` -/
+
+variable {E : Type*} [NormedAddCommGroup E]
+
+theorem integrable_traj {a b : ℕ} (hab : a ≤ b) {f : (Π n, X n) → E}
+    (x₀ : Π i : Iic a, X i) (i_f : Integrable f (traj κ a x₀)) :
+    ∀ᵐ x ∂traj κ a x₀, Integrable f (traj κ b (frestrictLe b x)) := by
+  rw [← traj_comp_partialTraj _ hab, integrable_comp_iff] at i_f
+  · apply ae_of_ae_map (p := fun x ↦ Integrable f (traj κ b x))
+    · fun_prop
+    · convert i_f.1
+      rw [← traj_map_frestrictLe, Kernel.map_apply _ (measurable_frestrictLe _)]
+  · exact i_f.aestronglyMeasurable
+
+theorem aestronglyMeasurable_traj {a b : ℕ} (hab : a ≤ b) {f : (Π n, X n) → E}
+    {x₀ : Π i : Iic a, X i} (hf : AEStronglyMeasurable f (traj κ a x₀)) :
+    ∀ᵐ x ∂partialTraj κ a b x₀, AEStronglyMeasurable f (traj κ b x) := by
+  rw [← traj_comp_partialTraj κ hab] at hf
+  exact hf.comp
+
+variable [NormedSpace ℝ E]
+
+variable {κ} in
+/-- When computing `∫ x, f x ∂traj κ n x₀`, because the trajectory up to time `n` is
+determined by `x₀` we can replace `x` by `updateFinset x _ x₀`. -/
+theorem integral_traj {a : ℕ} (x₀ : Π i : Iic a, X i) {f : (Π n, X n) → E}
+    (mf : AEStronglyMeasurable f (traj κ a x₀)) :
+    ∫ x, f x ∂traj κ a x₀ = ∫ x, f (updateFinset x _ x₀) ∂traj κ a x₀ := by
```

Comment:
Let's make that more readable:
```suggestion
/-- When computing `∫ x, f x ∂traj κ n x₀`, because the trajectory up to time `n` is
determined by `x₀` we can replace `x` by `updateFinset x (Iic a) x₀`. -/
theorem integral_traj {a : ℕ} (x₀ : Π i : Iic a, X i) {f : (Π n, X n) → E}
    (mf : AEStronglyMeasurable f (traj κ a x₀)) :
    ∫ x, f x ∂traj κ a x₀ = ∫ x, f (updateFinset x (Iic a) x₀) ∂traj κ a x₀ := by
```
And same suggestion in every statement where those appear. It's fine if you omit the finset in proofs.